src/main/java/appeng/me/storage/CellInventoryBase.java

```diff
@@ -0,0 +1,374 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Applied Energistics 2 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Applied Energistics 2.  If not, see <http://www.gnu.org/licenses/lgpl>.
+ */
+
+package appeng.me.storage;
+
+
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraftforge.items.IItemHandler;
+
+import appeng.api.config.FuzzyMode;
+import appeng.api.exceptions.AppEngException;
+import appeng.api.implementations.items.IStorageCell;
+import appeng.api.storage.ICellInventory;
+import appeng.api.storage.IMEInventory;
+import appeng.api.storage.ISaveProvider;
+import appeng.api.storage.data.IAEStack;
+import appeng.api.storage.data.IItemList;
+import appeng.util.Platform;
+
+
+/**
+ * @author DrummerMC
+ * @version rv6 - 2018-01-17
+ * @since rv6 2018-01-17
+ */
+public abstract class CellInventoryBase<T extends IAEStack<T>> implements ICellInventory<T>
+{
+
+	private static final String ITEM_TYPE_TAG = "it";
+	private static final String ITEM_COUNT_TAG = "ic";
+	private static final String ITEM_SLOT = "#";
+	private static final String ITEM_SLOT_COUNT = "@";
+	protected static final String ITEM_PRE_FORMATTED_COUNT = "PF";
+	protected static final String ITEM_PRE_FORMATTED_SLOT = "PF#";
+	protected static final String ITEM_PRE_FORMATTED_NAME = "PN";
+	protected static final String ITEM_PRE_FORMATTED_FUZZY = "FP";
+	private static String[] itemSlots;
+	private static String[] itemSlotCount;
+	private final NBTTagCompound tagCompound;
+	protected final ISaveProvider container;
+	private int maxItemTypes = 63;
+	private short storedItems = 0;
+	private int storedItemCount = 0;
+	protected IItemList<T> cellItems;
+	protected ItemStack i;
+	protected IStorageCell cellType;
+	private final int itemsPerByte;
+
+	protected CellInventoryBase( final NBTTagCompound data, final ISaveProvider container, final int itemsPerByte )
+	{
+		this.tagCompound = data;
+		this.container = container;
+		this.itemsPerByte = itemsPerByte;
+	}
+
+	protected CellInventoryBase( final ItemStack o, final ISaveProvider container, final int itemsPerByte ) throws AppEngException
+	{
+		this.itemsPerByte = itemsPerByte;
+		if( itemSlots == null )
+		{
+			itemSlots = new String[this.maxItemTypes];
+			itemSlotCount = new String[this.maxItemTypes];
+
+			for( int x = 0; x < this.maxItemTypes; x++ )
+			{
+				itemSlots[x] = ITEM_SLOT + x;
+				itemSlotCount[x] = ITEM_SLOT_COUNT + x;
+			}
+		}
+
+		if( o == null )
+		{
+			throw new AppEngException( "ItemStack was used as a cell, but was not a cell!" );
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "ItemStack was used as a cell, but was not a cell!" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1192)